tunnel/intra/doh/client_auth.go

```diff
@@ -0,0 +1,147 @@
+// Copyright 2020 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package doh
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"io"
+	"sync"
+
+	"github.com/eycorsican/go-tun2socks/common/log"
+)
+
+// CertificateLoader interface for requesting ClientAuth instances.
+type CertificateLoader interface {
+	// Request a ClientAuth instance (blocking).
+	// returns nil (no authentication) or a ClientAuth instance.
+	LoadClientCertificate() ClientAuth
+}
+
+// ClientAuth interface for providing TLS certificates and signatures.
+type ClientAuth interface {
+	// GetClientCertificate returns the client certificate (if any).
+	// It does not block or cause certificates to load.
+	// Returns a DER encoded X.509 client certificate.
+	GetClientCertificate() []byte
+	// GetIntermediateCertificate returns the chaining certificate (if any).
+	// It does not block or cause certificates to load.
+	// Returns a DER encoded X.509 certificate.
+	GetIntermediateCertificate() []byte
+	// Request a signature on a digest.
+	Sign(digest []byte) []byte
+}
+
+// clientAuthWrapper manages certificate loading and usage during TLS handshakes.
+// Implements crypto.Signer.
+type clientAuthWrapper struct {
+	loadCertificateOnce sync.Once
+	loader              CertificateLoader
+	signer              ClientAuth
```

Comment:
It looks like you're caching the signer here.  Is this a performance optimization, or is there a UX reason why this needs to be structured this way?  If caching is needed, I would prefer to either (1) push it into the Java code or (2) document the restriction on the CertificateLoader interface.

Reply:
I've removed all of the caching, we can reconsider if this overly complicates the Java code.